crates/wasi/src/preview2/preview1/mod.rs

```diff
@@ -1373,7 +1389,106 @@ impl<
         buf_len: types::Size,
         cookie: types::Dircookie,
     ) -> Result<types::Size, types::Error> {
-        todo!()
+        let fd = self.get_dir_fd(fd).await?;
+        let stream = self.read_directory(fd).await.map_err(|e| {
+            e.try_into()
+                .context("failed to call `read-directory`")
+                .unwrap_or_else(types::Error::trap)
+        })?;
+        let wasi::filesystem::DescriptorStat {
+            inode: fd_inode, ..
+        } = self.stat(fd).await.map_err(|e| {
+            e.try_into()
+                .context("failed to call `stat`")
+                .unwrap_or_else(types::Error::trap)
+        })?;
+        let cookie = cookie.try_into().map_err(|_| types::Errno::Overflow)?;
+
+        let head = [
+            (
+                types::Dirent {
+                    d_next: 1u64.to_le(),
+                    d_ino: fd_inode.to_le(),
+                    d_type: types::Filetype::Directory,
+                    d_namlen: 1u32.to_le(),
+                },
+                ".".into(),
+            ),
+            (
+                types::Dirent {
+                    d_next: 2u64.to_le(),
+                    d_ino: fd_inode.to_le(), // NOTE: legacy implementation returns `fd` inode here
```

Comment:
That is unfortunate, I never knew about this bug.

I am not sure what to do with this. We decided to get rid of providing `..` in the entries stream in preview 2, because it doesn't really fit the capability model - a preopen doesn't have a `..`. I don't think we have any way to implement this correctly, so I guess we just have to hope that this bug doesn't cause problems for userland